doc/man/man1/qhost.1
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004 by Sun Microsystems, Inc.
.\"
.\"___INFO__MARK_END__
.\"
.\" $RCSfile: qhost.1,v $ 
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\" "
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\"
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH QHOST 1 "2011-12-04" "xxRELxx" "xxQS_NAMExx User Commands"
.SH NAME
qhost \- show the status of xxQS_NAMExx hosts, queues, jobs
.\"
.\"
.SH SYNTAX
.B qhost
.RB [ \-F
.RI [ resource_name \fB,\fP...]]
.RB [ \-help ]
.RB [ \-h
.IR host_list ]
.RB [ \-j ]
.RB [ \-l
.IR resource [\fB=\fP val ]\fB,\fP...]
.RB [ \-ncb ]
.RB [ \-u
.IR user \fB,\fP...]
.RB [ \-xml ]
.\"
.SH DESCRIPTION
.I qhost
shows the current status of the available xxQS_NAMExx hosts, queues and the
jobs associated with the queues. Selection options allow you
to get information about specific hosts, queues, jobs or users.
If multiple selections are done a host is only displayed if all
selection criteria for a host are met.
Without any options
.I qhost
will display a list of all hosts without queue or job
information.
.PP
.\"
.SH OPTIONS
.\"
.IP "\fB\-F\fP [\fIresource_name\fP\fB,\fP...]"
.I qhost
will present a detailed listing of the current 
resource availability per host with respect to all resources (if the option 
argument is omitted) or with respect to those resources contained in the 
resource_name list. Please refer to the description of the
\fBFull Format\fP in 
section \fBOUTPUT FORMATS\fP below for further detail.
.\"
.IP "\fB\-help\fP"
Prints a listing of all options.
.\"
.IP "\fB\-h \fP\fIhost_list\fP"
Prints a list of all hosts contained in host_list.
.\"
.IP "\fB\-j\fP"
Prints all jobs running on the queues hosted by the shown hosts. This 
switch calls \fB\-q\fP implicitly.
.\"
.IP "\fB\-l \fP\fIresource\fP[\fB=\fP\fIvalue\fP]\fB,\fP..."
Defines the resources to be granted by the hosts which should be 
included in the host list output. Matching is performed on 
hosts based on non-mutable resource availability information only. 
That means load values are always ignored except the so-called static 
load values listed in
.M load_parameters 5 .
Also consumable utilization is ignored.
If there are multiple \-l resource requests they will be concatenated by
a logical AND: a host needs to match all resources to be displayed.
.\"
.IP "\fB\-ncb\fP"
This command line switch can be used in order to get 6.2u5 compatible
output with other
.M qhost 1
command line switches. In that case the output
of the corresponding command will suppress information concerning the
execution host topology. Note that this option will be removed in the
next major version.
.\"
.IP "\fB\-q\fP"
Show information about the queues instances hosted by the displayed hosts.
.\"
.IP "\fB\-u \fP\fIuser\fP\fB,\fP..."
Display information only on those jobs and queues
being associated with the users from the given user list.
.\"
.IP "\fB\-xml\fP"
This option can be used with all other options and changes the output to XML. The used
schemas are referenced in the XML output. The output is printed to stdout.
.sp
If the \fB\-xml\fP parameter is combined with \fB\-ncb\fP then the XML output will
contain 6.2u5 compatible output.
.br
.\"
.\"
.SH "OUTPUT FORMATS"
Depending on the presence or absence of the \fB\-q\fP or \fB\-F\fP and
\fB\-j\fP option three output formats need to be differentiated.
.\"
.SS "\fBDefault Format (without \-q, \-F and \-j)\fP"
For each host one line is printed. The output consists of
consisting of
.IP "\(bu" 3n
the Hostname 
.IP "\(bu" 3n
the Architecture.
.IP "\(bu" 3n
the  Number of processors.
.IP "\(bu" 3n
the Load.
.IP "\(bu" 3n
the Total Memory.
.IP "\(bu" 3n
the Used Memory.
.IP "\(bu" 3n
the Total Swapspace.
.IP "\(bu" 3n
the Used Swapspace.
.PP
More details can be found in
.M load_parameters 5 .
.\"
.PP
If the \fB\-q\fP option is supplied, each host status line also contains
extra lines for every queue hosted by the host consisting of,
.IP "\(bu" 3n
the queue name,
.IP "\(bu" 3n
the queue type \- one of B(atch), I(nteractive), or a combination thereof,
.IP "\(bu" 3n
the number of reserved (by advance reservation, not resource
reservation), used and, available (according to the queue's
.BR slots
parameter) job slots,
.IP "\(bu" 3n
the state of the queue \- one of
u(nknown), a(larm), A(larm), C(alendar suspended), s(uspended),
S(ubordinate), d(isabled), D(isabled), E(rror), (c)configuration
ambiguous, (o)rphaned, (P)reempted, or some combinations thereof.
See section "Full Format" in
.M qstat 1
for explanation of the status flags.
.PP
If the \fB\-F\fP option was used, resource availability information is printed 
following the host status line. For each resource (as selected in an option 
argument to \fB\-F\fP or for all resources if the option argument was
omitted) a single line is displayed with the following format:
.IP "\(bu" 3n
a one letter specifier indicating whether the current resource availability 
value was dominated by either
.br
`\fBg\fP' \- a cluster global,
.br
`\fBh\fP' \- a host total or
.IP "\(bu" 3n
a second one letter specifier indicating the source for the current resource 
availability value, being one of
.br
`\fBl\fP' \- a load value reported for the
resource,
.br
`\fBL\fP' \- a load value for the resource after administrator

NEWS
Notable changes in Son of Grid Engine
=====================================

For earlier changes (comprising Sun Grid Engine), see Changelog in the
source distribution, and the bug lists at
http://arc.liv.ac.uk/SGE/howto/Installation,%20Upgrade,%20Patches.
For detailed recent change information, including credits for changes
not listed here, see https://arc.liv.ac.uk/trac/SGE/log/sge/

"[#<n>]" below refers to ticket <n> at
https://arc.liv.ac.uk/trac/SGE/ticket/<n>.  For historical reasons,
changes due to Univa/Sun are tagged as "[U]", or "[(U)]" if partially
due to them or appearing first in this version.


Version 8.1.9
-------------

* Security
  * Drop setgid privileges correctly (CERT POS36-C); daemons now run
    with (only) admin user's groups, not root's

* Building

  * MUNGE authentication can be built in (aimk -munge)
  * Use a minimal autoconf framework
  * Fix building on recent Darwin and Cygwin [#1516, #1536, #1557]
  * Drop support for Darwin < 10 and Darwin/PPC
  * Cope with GNU/Linux ppc64le as well as ppc64el
  * Fix rpm packaging of spool objects [#1519]
  * Actually install "work" in utilbin (c.f. 8.1.8)
  * Rpm spec fixes for Fedora 23, ppc64, and OpenSuSE
  * Fix build when bool is defined in headers

* Bug fixes

  * Man fixes
  * Message fixes [#1538, #1539]
  * Fix RPM packaging of spool files [#1519]
  * Don't set port and cluster in check_sge.py, and accept memory in T
  * Fix pam_sge-qrsh-setup for recent PAM
  * Do a full, not a partial match on GECOS data when selecting certificate
    in sge_ca [#1517]
  * Fix checking of admin host in qconf [#1547]
  * Report the actual exit status of master task in tightly integrated
    parallel jobs
  * Fix share tree accounting with array jobs [#435]
  * Fix doubled multipliers of units in qacct memory reports
  * Improve job scheduling within a sharetree node [#1550]
  * Flush spool at qmaster exit [#1551]
  * Sort jobs and advance reservations when reading from spool [#1552, #1553]
  * Fix detection of old v. new type of cpuset [#1523]
  * Fix treatment of some invalid entries in variable lists or the
    environment on job submission

* Enhancements

  * MUNGE authentication (as above)
  * qconf -clearusage can act on a specific user
  * Allow generating load sensor data just before sending them
  * Add sge_ca "-revoke <cert>" option
  * Allow running DRMAA jobs as arbitrary user [#492].  (Partially
    done in v 8.1.0, but not mentioned; now works with CSP.)
  * Report SSL error numbers in hex [#1539]
  * Use PVM_VMID environment variable [#44]
  * rc templates support a "status" argument (for HA usage)

* Potentially incompatible changes

  * Print error messages to stderr and don't print on success with -terse
    to be consistent with submit(1) saying that error messages go to

source/clients/qhost/qhost.c
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 * 
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 * 
 *  Sun Microsystems Inc., March, 2001
 * 
 * 
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 * 
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 * 
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 * 
 *   Copyright: 2001 by Sun Microsystems, Inc.
 * 
 *   All Rights Reserved.
 *
 *  Portions of this code are Copyright 2011 Univa Corporation.
 *
 ************************************************************************/
/*___INFO__MARK_END__*/
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <math.h>

#include "uti/sge_rmon.h"
#include "uti/sge_bootstrap.h"
#include "uti/sge_prog.h"
#include "uti/sge_unistd.h"
#include "uti/sge_stdlib.h"
#include "uti/sge_hostname.h"
#include "uti/sge_log.h"
#include "uti/sge_profiling.h"

#include "sgeobj/sge_all_listsL.h"
#include "sgeobj/sge_feature.h"
#include "sgeobj/parse.h"
#include "sgeobj/sge_host.h"
#include "sgeobj/sge_range.h"
#include "sgeobj/sge_answer.h"
#include "sgeobj/sge_centry.h"
#include "sgeobj/sge_cull_xml.h"

#include "gdi/sge_gdi.h"
#include "gdi/sge_gdi_ctx.h"

#include "sched/sge_select_queue.h"

#include "comm/commlib.h"

#include "sge.h"
#include "basis_types.h"
#include "sig_handlers.h"
#include "qstat_printing.h"
#include "sge_qhost.h"
#include "msg_common.h"
#include "msg_clients_common.h"
#include "msg_qhost.h"

extern char **environ;

static bool sge_parse_cmdline_qhost(char **argv, char **envp, lList **ppcmdline, lList **alpp);
static int sge_parse_qhost(lList **ppcmdline, lList **pplres, lList **ppFres, lList **pphost, lList **ppuser, u_long32 *show, qhost_report_handler_t **report_handler, lList **alpp);
static bool qhost_usage(FILE *fp);

static qhost_report_handler_t* xml_report_handler_create(lList **alpp);
static int xml_report_handler_destroy(qhost_report_handler_t** handler, lList **alpp);
static int xml_report_finished(qhost_report_handler_t* handler, lList **alpp);
static int xml_report_started(qhost_report_handler_t* handler, lList **alpp);
static int xml_report_host_begin(qhost_report_handler_t* handler, const char* host_name, lList **alpp);
static int xml_report_host_string_value(qhost_report_handler_t* handler, const char* name, const char *value, lList **alpp);
static int xml_report_host_ulong_value(qhost_report_handler_t* handler, const char* name, u_long32 value, lList **alpp);
static int xml_report_host_finished(qhost_report_handler_t* handler, const char* host_name, lList **alpp);
static int xml_report_resource_value(qhost_report_handler_t* handler, const char* dominance, const char* name, const char* value, lList **alpp);
static int xml_report_queue_begin(qhost_report_handler_t* handler, const char* qname, lList **alpp);
static int xml_report_queue_string_value(qhost_report_handler_t* handler, const char* qname, const char* name, const char *value, lList **alpp);
static int xml_report_queue_ulong_value(qhost_report_handler_t* handler, const char* qname, const char* name, u_long32 value, lList **alpp);
static int xml_report_queue_finished(qhost_report_handler_t* handler, const char* qname, lList **alpp);
static int xml_report_job_begin(qhost_report_handler_t* handler, const char *qname, const char* jobname, lList **alpp);
static int xml_report_job_string_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, const char *value, lList **alpp);
static int xml_report_job_ulong_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, u_long32 value, lList **alpp);
static int xml_report_job_double_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, double value, lList **alpp);
static int xml_report_job_finished(qhost_report_handler_t* handler, const char *qname, const char* jobname, lList **alpp);




static int xml_report_started(qhost_report_handler_t* handler, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_started");

   printf("<?xml version='1.0'?>\n");
   printf("<qhost xmlns:xsd=\"http://arc.liv.ac.uk/repos/darcs/sge/source/dist/util/resources/schemas/qhost/qhost.xsd\">\n");
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_finished(qhost_report_handler_t* handler, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_finished");
   
   printf("</qhost>\n");
   
   DRETURN(QHOST_SUCCESS);
}

static qhost_report_handler_t* xml_report_handler_create(lList **alpp)
{
   qhost_report_handler_t* ret = (qhost_report_handler_t*)sge_malloc(sizeof(qhost_report_handler_t));

   DENTER(TOP_LAYER, "xml_report_handler_create");

   if (ret == NULL ) {
      answer_list_add_sprintf(alpp, STATUS_EMALLOC, ANSWER_QUALITY_ERROR,
                              MSG_MEM_MEMORYALLOCFAILED_S, SGE_FUNC);
      DRETURN(NULL);
   }
   /*
   ** for xml_report_handler ctx is a dstring
   */
   ret->ctx = sge_malloc(sizeof(dstring));
   if (ret->ctx == NULL ) {
      answer_list_add_sprintf(alpp, STATUS_EMALLOC, ANSWER_QUALITY_ERROR,
                              MSG_MEM_MEMORYALLOCFAILED_S, SGE_FUNC);
      DRETURN(NULL);
   }
   /*
   ** corresponds to initializing with DSTRING_INIT
   */
   memset(ret->ctx, 0, sizeof(dstring));
   
   ret->report_started = xml_report_started;
   ret->report_finished = xml_report_finished;
   ret->report_host_begin = xml_report_host_begin;
   ret->report_host_string_value = xml_report_host_string_value;
   ret->report_host_ulong_value = xml_report_host_ulong_value;
   ret->report_host_finished = xml_report_host_finished;
   
   ret->report_resource_value = xml_report_resource_value;

   ret->report_queue_begin = xml_report_queue_begin;
   ret->report_queue_string_value = xml_report_queue_string_value;
   ret->report_queue_ulong_value = xml_report_queue_ulong_value;
   ret->report_queue_finished = xml_report_queue_finished;
   
   ret->report_job_begin = xml_report_job_begin;
   ret->report_job_string_value = xml_report_job_string_value;
   ret->report_job_ulong_value = xml_report_job_ulong_value;
   ret->report_job_double_value = xml_report_job_double_value;
   ret->report_job_finished = xml_report_job_finished;

   ret->destroy = xml_report_handler_destroy;

   DRETURN(ret);
}

static int xml_report_handler_destroy(qhost_report_handler_t** handler, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_handler_destroy");

   if (handler != NULL && *handler != NULL ) {
      sge_dstring_free((dstring*)(*handler)->ctx);
      sge_free(&((*handler)->ctx));
      sge_free(handler);
      *handler = NULL;
   }

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_host_begin(qhost_report_handler_t* handler, const char* host_name, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_host_begin");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(host_name, (dstring*)handler->ctx);
   printf(" <host name='%s'>\n", sge_dstring_get_string((dstring*)handler->ctx));

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_host_string_value(qhost_report_handler_t* handler, const char*name, const char *value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_host_string_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);
   printf("   <hostvalue name='%s'>", sge_dstring_get_string((dstring*)handler->ctx) );
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(value, (dstring*)handler->ctx);
   printf("%s</hostvalue>\n", sge_dstring_get_string((dstring*)handler->ctx));

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_host_ulong_value(qhost_report_handler_t* handler, const char* name, u_long32 value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_host_ulong_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);
   printf("   <hostvalue name='%s'>"sge_U32CFormat"</hostvalue>\n", sge_dstring_get_string((dstring*)handler->ctx), sge_u32c(value));

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_host_finished(qhost_report_handler_t* handler, const char* host_name, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_host_finished");

   printf(" </host>\n");   

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_resource_value(qhost_report_handler_t* handler, const char* dominance, const char* name, const char* value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_resource_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);
   printf("   <resourcevalue name='%s' ", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(dominance, (dstring*)handler->ctx);   
   printf("dominance='%s'>", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(value, (dstring*)handler->ctx);   
   printf("%s</resourcevalue>\n", sge_dstring_get_string((dstring*)handler->ctx));
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_queue_begin(qhost_report_handler_t* handler, const char* qname, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_queue_begin");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(qname, (dstring*)handler->ctx);
   printf(" <queue name='%s'>\n", sge_dstring_get_string((dstring*)handler->ctx));

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_queue_string_value(qhost_report_handler_t* handler, const char* qname, const char* name, const char *value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_queue_string_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(qname, (dstring*)handler->ctx);
   printf("   <queuevalue qname='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);      
   printf(" name='%s'>", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(value, (dstring*)handler->ctx);      
   printf("%s</queuevalue>\n", sge_dstring_get_string((dstring*)handler->ctx));
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_queue_ulong_value(qhost_report_handler_t* handler, const char* qname, const char* name, u_long32 value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_queue_ulong_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(qname, (dstring*)handler->ctx);
   printf("   <queuevalue qname='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);      
   printf(" name='%s'>"sge_U32CFormat"</queuevalue>\n", sge_dstring_get_string((dstring*)handler->ctx), sge_u32c(value));
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_queue_finished(qhost_report_handler_t* handler, const char* qname, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_queue_finished");

   printf(" </queue>\n");   

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_job_begin(qhost_report_handler_t* handler, const char *qname, const char* jobname, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_job_begin");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(jobname, (dstring*)handler->ctx);
   printf(" <job name='%s'>\n", sge_dstring_get_string((dstring*)handler->ctx));

   DRETURN(QHOST_SUCCESS);
}

static int xml_report_job_string_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, const char *value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_job_string_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(jobname, (dstring*)handler->ctx);
   printf("   <jobvalue jobid='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);      
   printf(" name='%s'>", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(value, (dstring*)handler->ctx);      
   printf("%s</jobvalue>\n", sge_dstring_get_string((dstring*)handler->ctx));
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_job_ulong_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, u_long32 value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_job_ulong_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(jobname, (dstring*)handler->ctx);
   printf("   <jobvalue jobid='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);      
   printf(" name='%s'>"sge_U32CFormat"</jobvalue>\n", sge_dstring_get_string((dstring*)handler->ctx), sge_u32c(value));
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_job_double_value(qhost_report_handler_t* handler, const char *qname, const char* jobname, const char* name, double value, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_job_double_value");

   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(jobname, (dstring*)handler->ctx);
   printf("   <jobvalue jobid='%s'", sge_dstring_get_string((dstring*)handler->ctx));
   
   sge_dstring_clear((dstring*)handler->ctx);
   escape_string(name, (dstring*)handler->ctx);      
   printf(" name='%s'>'%f'</jobvalue>\n", sge_dstring_get_string((dstring*)handler->ctx), value);
   
   DRETURN(QHOST_SUCCESS);
}

static int xml_report_job_finished(qhost_report_handler_t* handler, const char *qname, const char* job_name, lList **alpp)
{
   DENTER(TOP_LAYER, "xml_report_job_finished");

   printf(" </job>\n");   

   DRETURN(QHOST_SUCCESS);
}

                                      
/************************************************************************/
int main(int argc, char **argv)
{
   lList *pcmdline = NULL;
   lList *ul = NULL;
   lList *host_list = NULL;
   u_long32 show = 0;
   lList *resource_list = NULL;
   lList *resource_match_list = NULL;
   lList *alp = NULL;
   qhost_report_handler_t *report_handler = NULL;
   int is_ok = 0;
   int qhost_result = 0;
   sge_gdi_ctx_class_t *ctx = NULL;

   DENTER_MAIN(TOP_LAYER, "qhost");

   sge_sig_handler_in_main_loop = 0;
   log_state_set_log_gui(true);
   sge_setup_sig_handlers(QHOST);

   if (sge_gdi2_setup(&ctx, QHOST, MAIN_THREAD, &alp) != AE_OK) {
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT((void**)&ctx, 1);
   }

   /*
   ** stage 1 of commandline parsing
   */
   if (!sge_parse_cmdline_qhost(argv, environ, &pcmdline, &alp)) {
      /*
      ** high level parsing error! sow answer list
      */
      answer_list_output(&alp);
      lFreeList(&pcmdline);
      sge_prof_cleanup();
      SGE_EXIT((void **)&ctx, 1);
   }

   /*
   ** stage 2 of commandline parsing 
   */
   is_ok = sge_parse_qhost(&pcmdline, 
                           &resource_match_list,   /* -l resource_request           */
                           &resource_list,         /* -F qresource_request          */
                           &host_list,             /* -h host_list                  */
                           &ul,                    /* -u user_list                  */
                           &show,                  /* -q, -j                        */
                           &report_handler,
                           &alp);
   lFreeList(&pcmdline);
   if (is_ok == 0) {     
      /*
      ** low level parsing error! show answer list
      */
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 1);
   } else if (is_ok == 2) {
      /* -help output generated, exit normally */ 
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT(NULL, 0);
   }

   qhost_result = do_qhost(ctx, host_list, ul, resource_match_list, resource_list, 
                           show, &alp, report_handler);

   if (report_handler != NULL) {
      report_handler->destroy(&report_handler, &alp);
   }
   
   if (qhost_result != QHOST_SUCCESS) {
      answer_list_output(&alp);
      sge_prof_cleanup();
      SGE_EXIT((void**)&ctx, 1);
   }

   sge_prof_cleanup();
   SGE_EXIT((void**)&ctx, 0); /* 0 means ok - others are errors */
   DEXIT;
   return 0;
}


/*
** NAME
**   qhost_usage
** PARAMETER
**   none
** RETURN
**   none
** EXTERNAL
**   none
** DESCRIPTION
**   displays qhost_usage for qlist client
**   note that the other clients use a common function
**   for this. output was adapted to a similar look.
*/
static bool qhost_usage(
FILE *fp 
) {
   dstring ds;
   char buffer[256];

   DENTER(TOP_LAYER, "qhost_usage");

   if (fp == NULL) {
      DRETURN(false);
   }

   sge_dstring_init(&ds, buffer, sizeof(buffer));

   fprintf(fp, "%s\n", feature_get_product_name(FS_SHORT_VERSION, &ds));

   fprintf(fp,"%s qhost [options]\n", MSG_SRC_USAGE);
         
   fprintf(fp, "  [-F [resource_attribute]]  %s\n", MSG_QHOST_F_OPT_USAGE); 
   fprintf(fp, "  [-h hostlist]              %s\n", MSG_QHOST_h_OPT_USAGE);
   fprintf(fp, "  [-help]                    %s\n", MSG_COMMON_help_OPT_USAGE);
   fprintf(fp, "  [-j]                       %s\n", MSG_QHOST_j_OPT_USAGE);
   fprintf(fp, "  [-l attr=val,...]          %s\n", MSG_QHOST_l_OPT_USAGE);
   fprintf(fp, "  [-ncb]                     %s\n", MSG_QHOST_ncb_OPT_USAGE);
   fprintf(fp, "  [-q]                       %s\n", MSG_QHOST_q_OPT_USAGE);
   fprintf(fp, "  [-u user[,user,...]]       %s\n", MSG_QHOST_u_OPT_USAGE); 
   fprintf(fp, "  [-xml]                     %s\n", MSG_COMMON_xml_OPT_USAGE);

   DRETURN(true);
}

/****
 **** sge_parse_cmdline_qhost (static)
 ****
 **** 'stage 1' parsing of qhost-options. Parses options
 **** with their arguments and stores them in ppcmdline.
 ****/ 
static bool sge_parse_cmdline_qhost(
char **argv,
char **envp,
lList **ppcmdline,
lList **alpp
) {
   char **sp;
   char **rp;
   DENTER(TOP_LAYER, "sge_parse_cmdline_qhost");

   rp = ++argv;

   while(*(sp=rp)) {
      /* -help */
      if ((rp = parse_noopt(sp, "-help", NULL, ppcmdline, alpp)) != sp)
         continue;
 
      /* -ncb */
      if ((rp = parse_noopt(sp, "-ncb", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -q */
      if ((rp = parse_noopt(sp, "-q", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -F */
      if ((rp = parse_until_next_opt2(sp, "-F", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -h */
      if ((rp = parse_until_next_opt(sp, "-h", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -j */
      if ((rp = parse_noopt(sp, "-j", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -l */
      if ((rp = parse_until_next_opt(sp, "-l", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -u */
      if ((rp = parse_until_next_opt(sp, "-u", NULL, ppcmdline, alpp)) != sp)
         continue;

      /* -xml */
      if ((rp = parse_noopt(sp, "-xml", NULL, ppcmdline, alpp)) != sp)
         continue;
      
      /* oops */
      qhost_usage(stderr);
      answer_list_add_sprintf(alpp, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR, MSG_PARSE_INVALIDOPTIONARGUMENTX_S, *sp);
      DRETURN(false);
   }
   DRETURN(true);
}

/****
 **** sge_parse_qhost (static)
 ****
 **** 'stage 2' parsing of qhost-options. Gets the options from pcmdline
 ****/
static int sge_parse_qhost(lList **ppcmdline,
                            lList **pplres,
                            lList **ppFres,
                            lList **pphost,
                            lList **ppuser,
                            u_long32 *show,
                            qhost_report_handler_t **report_handler,
                            lList **alpp) 
{
   u_long32 helpflag;
   bool usageshowed = false;
   u_long32 full = 0;
   u_long32 binding = 0;
   char * argstr = NULL;
   lListElem *ep;
   int ret = 1;

   DENTER(TOP_LAYER, "sge_parse_host");
 
   /* Loop over all options. Only valid options can be in the
      ppcmdline list. 
   */

   /* display topology related information per default */
   (*show) |= QHOST_DISPLAY_BINDING;

   while (lGetNumberOfElem(*ppcmdline))
   {
      if (parse_flag(ppcmdline, "-help",  alpp, &helpflag)) {
         usageshowed = true;
         qhost_usage(stdout);
         ret = 2;
         goto exit;   
      }

      if (parse_multi_stringlist(ppcmdline, "-h", alpp, pphost, ST_Type, ST_name)) {
         /* 
         ** resolve hostnames and replace them in list
         */
         for_each(ep, *pphost) {
            if (sge_resolve_host(ep, ST_name) != CL_RETVAL_OK) {
               char buf[BUFSIZ];
               sprintf(buf, MSG_SGETEXT_CANTRESOLVEHOST_S, lGetString(ep,ST_name) );
               answer_list_add(alpp, buf, STATUS_ESYNTAX, ANSWER_QUALITY_ERROR);
               goto error;
            }
         }
         continue;
      }

      if (parse_multi_stringlist(ppcmdline, "-F", alpp, ppFres, ST_Type, ST_name)) {
         (*show) |= QHOST_DISPLAY_RESOURCES;
         continue;
      }
      if (parse_flag(ppcmdline, "-ncb", alpp, &binding)) {
         /* disable topology related information */
         (*show) ^= QHOST_DISPLAY_BINDING;
         continue;
      }
      if (parse_flag(ppcmdline, "-q", alpp, &full)) {
         if(full) {
            (*show) |= QHOST_DISPLAY_QUEUES;
            full = 0;
         }
         continue;
      }

      if (parse_flag(ppcmdline, "-j", alpp, &full)) {
         if(full) {
            (*show) |= QHOST_DISPLAY_JOBS;
            full = 0;
         }
         continue;
      }
      while (parse_string(ppcmdline, "-l", alpp, &argstr)) {
         *pplres = centry_list_parse_from_string(*pplres, argstr, false);
         sge_free(&argstr);
         continue;
      }
      
      if (parse_multi_stringlist(ppcmdline, "-u", alpp, ppuser, ST_Type, ST_name)) {
         (*show) |= QHOST_DISPLAY_JOBS;
         continue;
      }

      if (parse_flag(ppcmdline, "-xml", alpp, &full)) {
         *report_handler = xml_report_handler_create(alpp);
         continue;
      }

   }
   if (lGetNumberOfElem(*ppcmdline)) {
     answer_list_add_sprintf(alpp, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR,
                             "%s", MSG_PARSE_TOOMANYOPTIONS);
     goto error;
   }

   DRETURN(1);

   error:
      ret = 0;
   exit:
      if (!usageshowed) {
         qhost_usage(stderr);
      }
      if (report_handler && *report_handler) {
         (*report_handler)->destroy(report_handler, alpp);
      }
      lFreeList(pplres);
      lFreeList(ppFres);
      lFreeList(pphost);
      lFreeList(ppuser);
   
      DRETURN(ret);
}

